app/assets/javascripts/discourse/app/mixins/composer-upload-uppy.js

```diff
@@ -140,24 +140,37 @@ export default Mixin.create(ExtendableUploader, UppyS3Multipart, {
       },
 
       onBeforeUpload: (files) => {
-        const fileCount = Object.keys(files).length;
         const maxFiles = this.siteSettings.simultaneous_uploads;
 
         // Look for a matching file upload handler contributed from a plugin.
-        // It is not ideal that this only works for single file uploads, but
-        // at this time it is all we need. In future we may want to devise a
-        // nicer way of doing this. Uppy plugins are out of the question because
-        // there is no way to define which uploader plugin handles which file
-        // extensions at this time.
-        if (fileCount === 1) {
-          const file = Object.values(files)[0];
+        // In future we may want to devise a nicer way of doing this.
+        // Uppy plugins are out of the question because there is no way to
+        // define which uploader plugin handles which file extensions at this time.
+        const uploadHandlerFiles = [];
+        const unhandledFiles = {};
+        for (const [fileId, file] of Object.entries(files)) {
           const matchingHandler = this._findMatchingUploadHandler(file.name);
+          if (matchingHandler) {
+            uploadHandlerFiles.push([file, matchingHandler]);
+          } else {
+            unhandledFiles[fileId] = { ...files[fileId] };
+          }
+        }
+
+        // This replicates what the old composer upload handler did; because
+        // jQuery file uploader passed through a single file at a time to
+        // the upload handlers.
+        uploadHandlerFiles.forEach((fileWithHandler) => {
+          const file = fileWithHandler[0];
+          const matchingHandler = fileWithHandler[1];
```

Comment:
You could do the same nice destructuring here like you did on line 151 🙂

Reply:
How does my brain do this thing where it remembers a language feature 10 lines up but not here? :brain: :-1: